packages/artifact/src/internal/download-http-client.ts

```diff
@@ -124,66 +160,124 @@ export class DownloadHttpClient {
+          info(
+            `Backoff due to too many requests, retry #${retryCount}. Waiting for ${retryAfterValue} milliseconds before continuing the download`
+          )
+          await new Promise(resolve => setTimeout(resolve, retryAfterValue))
+        } else {
+          // Back off using an exponential value that depends on the retry count
+          const backoffTime = getExponentialRetryTimeInMilliseconds(retryCount)
+          info(
+            `Exponential backoff for retry #${retryCount}. Waiting for ${backoffTime} milliseconds before continuing the download`
+          )
+          await new Promise(resolve => setTimeout(resolve, backoffTime))
+        }
+        info(
+          `Finished backoff for retry #${retryCount}, continuing with download`
+        )
+      }
+    }
+
+    // keep trying to download a file until a retry limit has been reached
+    while (retryCount <= retryLimit) {
+      let response: IHttpClientResponse
+      try {
+        response = await makeDownloadRequest()
+      } catch (error) {
+        // if an error is caught, it is usually indicative of a timeout so retry the download
+        info('An error has been caught, while attempting to download a file')
+        // eslint-disable-next-line no-console
+        console.log(error)
+
+        // increment the retryCount and use exponential backoff to wait before making the next request
+        await backOff()
+        continue
+      }
+
+      if (isSuccessStatusCode(response.message.statusCode)) {
+        // The body contains the contents of the file however calling response.readBody() causes all the content to be converted to a string
+        // which can cause some gzip encoded data to be lost
+        // Instead of using response.readBody(), response.message is a readableStream that can be directly used to get the raw body contents
+        await this.pipeResponseToFile(
           response,
-          stream,
+          destinationStream,
           isGzip(response.message.headers)
         )
+        return
+      } else if (isRetryableStatusCode(response.message.statusCode)) {
+        info(
+          `A ${response.message.statusCode} response code has been received while attempting to download an artifact`
+        )
+        // if a throttled status code is received, try to get the retryAfter header value, else differ to standard exponential backoff
+        isThrottledStatusCode(response.message.statusCode)
+          ? await backOff(
+              tryGetRetryAfterValueTimeInMilliseconds(response.message.headers)
+            )
+          : await backOff()
       } else {
+        // Some unexpected response code, fail immediately and stop the download
         // eslint-disable-next-line no-console
-        console.log(retryResponse)
-        throw new Error(`Unable to download ${artifactLocation}`)
+        console.log(response)
+        throw new Error(
+          `###ERROR### Unable to download ${artifactLocation} ###`
```

Comment:
What's up with the `###ERROR###` here? If we want to log an error you can just use `core.Error` and then throw the normal message.
We should also be logging `response` as an error too.

Reply:
It's not so much about logging an error, but here we want to throw one so that execution halts. 

`core.error(response.toString())` is useless because it just outputs `[object Object]` which is why you see `console.log` being used in a few places.

I've also cleaned this up a little. Outputting the entire response if pretty useless and just pollutes the logs, the only thing we really care about is the status code and url that we're using for the request.